### PR TITLE
Feature/amd stack

### PR DESCRIPTION
**Pull request by BiG86**
# Add a llama.cpp backend so Rizzo Flow runs on AMD GPUs

Closes the gap where MLX is the only runtime: Radeon (and Intel, and anything else
llama.cpp drives) can now run the same typed-decision API. Prompt, decision layer and
HTTP contract are untouched — this is a second scoring backend behind the same surface.

## Why

MLX has no AMD backend. Until now the only options for a Radeon card were "CPU" (measured
at ~185 s for an 8-token forward pass of the 1.7B) or nothing. `spark2_5` is upstream in
llama.cpp since [PR #27868](https://github.com/ggml-org/llama.cpp/pull/27868), so the
conversion path exists and does not need a fork; this pins v0.4.1 (`b29c606`) and a
sha256-pinned GGUF.

While porting, the selection logic turned out to have three silent-wrong-answer bugs that
this PR also fixes (see *Hardware-aware selection* below).

## What this adds

**The runtime** — three modules, no changes above them:

| File | Role |
| --- | --- |
| `llama_runtime.py` | ctypes bindings to `libllama`: structs transcribed from the pinned header, flat batches (no padding), `llama_memory_seq_cp`/`_seq_rm` for KV branching, vocab, GGUF metadata, device resolution |
| `llama_tokenizer.py` | renders the GGUF's `chat_template.jinja` with `jinja2` and tokenizes through `llama_tokenize` |
| `backend_llama.py` | identity/fingerprint (two provenance layers: model from `config.MODELS`, artifact from `config.GGUF_MODELS`) and `score()` |

`score()` implements both modes: `shared` prefills the state once and clones the KV cache
per question, `direct` rebuilds it per job. The suffix micro-batches are sized by
`--batch-size`.

**Pinned artifacts** — `stornic56/Spark-X2.5-4B-GGUF` at revision `7ce72e5c…`; `q8_0`
(4.4 GB) and `bf16` (8.2 GB) with sha256 verified before load, and re-checked at load.

**Hardware-aware selection** — `hardware.py` (new) probes the accelerators actually
present (Linux DRM PCI ids, NVIDIA driver nodes, Apple Silicon) and `backends.route()`
combines that with per-stack capabilities. Because of this:

- `mlx-cpu` can no longer outrank a usable llama.cpp. It imports on any machine, so the
  old `auto` picked it first and cost minutes per decision on an AMD box while Vulkan was
  available.
- A CUDA llama.cpp build is detected (`libggml-cuda.so`), so an NVIDIA box with that build
  no longer falls to CPU.
- Device vocabulary is hardware-centric: `auto|gpu|apple|nvidia|amd|intel|cpu`, with
  `mlx|vulkan|hip` kept as stack pins and `cuda` as an NVIDIA alias. A pin that cannot be
  honoured **raises**; only `auto` may fall back to CPU, and then it reports
  `downgraded=True` and warns.
- `gguf_path(quant, size)` refuses a pin of another size, so `--size 1.7b --quant q8_0`
  errors instead of silently loading the 4B.
- `llama_present()` uses this platform's library name, so Windows and macOS stop reporting
  "no llama backend".

**Docs** — `docs/llama-amd.md` (full runbook), README reorganised so the two runtimes are
first-class rather than "three extras plus an appendix", `CLAUDE.md`, and
`results/README.md`.

## What is deliberately not touched

`engine.py`, `prompts.py`, `decisions.py`, `schema.py`, `responses.py`, `compat.py`,
`playground.html`, `snake.html` and the MLX `backend.py`. The decision layer does not know
which stack is running, and the existing 87 tests for it pass unchanged.

## Design decisions worth review

1. **`--ctx` is per-sequence, and llama.cpp reserves rather than guards.** Prefix branching
   duplicates the KV cache, so `load()` requests `ctx × (batch_size + 1)` tokens and
   validates `n_ctx_seq ≥ ctx`. This is not a cosmetic detail: a `seq_cp` beyond
   `n_seq_max` **aborts the process** (a core dump during development), so the backend
   refuses the combination instead.
2. **MLX ranks before llama.cpp for the same accelerator** (so NVIDIA with the CUDA extra
   prefers MLX-CUDA). This is a documented default, **not a measurement** — no MLX-vs-llama
   figure exists on one GPU.
3. **On CPU, llama.cpp is preferred over MLX** (`CPU_STACK_PREFERENCE`), because MLX's CPU
   backend is documented as impractical in this very repo.
4. **A pinned stack with nothing installed defers to the loader** (`Routing(backend, "auto")`)
   so the error names the library path, which the router cannot know.

## Verification

```bash
uv sync --locked --extra llama --extra test
pytest -q                                  # 117 passed, 11 skipped
RIZZO_LLAMA_TEST=1 pytest -q -m llama      # 4 passed (loads the real 4 GB GGUF)
ruff check src tests scripts && ruff format --check src tests scripts
```

Weight-free unit tests grew from 87 to **117**; the four integration tests are opt-in and
are the only ones that touch real weights. The lib-facing paths use a fake library, so the
default suite stays at ~0.4 s.

Measured on Linux + RX 7900 XTX, Vulkan, `q8_0`, prompt v3 (committed report:
`results/llama-q8-vulkan-validation/validation.json`):

| Measurement | Value |
| --- | ---: |
| Smoke (17 rows) accuracy | 0.95 |
| Smoke NLL / Brier / ECE | 0.450 / 0.079 / 0.041 |
| Coverage / status accuracy | 0.80 / 0.95 |
| Throughput | 7.19 decisions/s |
| Latency p50 / p95 | 148 ms / 375 ms |
| Ticket: shared prefix / batches | 186 tokens / 2 |
| Ticket: shared vs direct | 0.327 s / 0.379 s, same argmax |
| Load (sha256 + model) | ~3.7 s cold, ~1.3 s warm |

Re-running the endpoint reproduced the committed numbers exactly (0.95 / 0.450 / 0.041) and
was deterministic across three identical requests. Prompt local proofs recorded in the
report: the GGUF template renders identically to the HuggingFace one, every letter A–Z is a
single token (46–71), and our tokenizer agrees with `llama-tokenize` from the same build.

## Honesty: what is not claimed

**This is not a cross-backend parity claim.** The MLX numbers in the README come from
different hardware and a different runtime; the smoke score coincidentally landing on the
same 0.95 is a sanity check, not equivalence. `prompt_sha256` and `input_tokens` are stored
per answer so a real comparison is possible later.

The NVIDIA routing paths are covered by injected hardware and capabilities only — no NVIDIA
GPU or CUDA/HIP build exists on the machine where this was developed. `detect_backends` does
not look for `ggml-metal`, so a Mac asking for `--gguf` routes to CPU (marked `NOTICED` in
the code).

## Known limitations

- No selected-token projection: llama.cpp projects the full vocabulary (the MLX stack
  multiplies only the answer-letter rows).
- The API cannot report peak memory, so the llama backend does not either.
- Calibrations are not portable across backends: the fingerprint includes the runtime.
- The `uv.lock` rewrite is from uv 0.12.17 (adds `upload-time` metadata); no dependency
  versions changed apart from the new `jinja2`/`markupsafe` pair.

## Reviewer map

| Order | File | Why |
| --- | --- | --- |
| 1 | `hardware.py` | smallest, self-contained; the probe contract |
| 2 | `backends.py` | the routing policy and the device vocabulary — the part most likely to need a decision from a maintainer |
| 3 | `llama_runtime.py` | struct layouts and the `n_ctx`/`seq_cp` invariants |
| 4 | `backend_llama.py` | identity/fingerprint and `score()` |
| 5 | `llama_tokenizer.py` | template rendering |
| 6 | tests + docs | behaviour pins and the honest write-up |

## Checklist

- [x] Tests added for every new behaviour (30 new unit tests, 3 integration)
- [x] `pytest -q` passes; integration suite passes with `RIZZO_LLAMA_TEST=1`
- [x] ruff check + format clean
- [x] No changes to the decision layer, the HTTP contract or the MLX backend
- [x] Numbers in README/docs match the committed report
- [x] Limitations documented in the README and in the code
- [ ] NVIDIA and Intel paths exercised on real hardware (needs a device)
- [ ] HIP build run (only Vulkan was built here)

## Follow-ups (not in this PR)

Selected-token projection for llama.cpp; a `hip` run; 1.7B GGUF pin; an OpenAI-compatible
or MCP surface; the SemIf comparison on the same GPU.

### CODE REVIEW

**Comment by simone-rizzo** [comment]
Thank you, @BiG86 — this pull request is the reason Rizzo Flow now runs on llama.cpp.

Until you opened it, MLX was the only runtime and an AMD card had no real way to run the project. Your work showed that the whole pattern (one forward pass, answer-letter logits, a shared state prefix branched per question) maps cleanly onto llama.cpp, that `spark2_5` is upstream, and your measurements on a Radeon showed it holds up there. That convinced us to go further than a second backend: as of [`40a21fd`](https://github.com/Rizzo-AI-Academy/rizzo-flow/commit/40a21fd) **llama.cpp is the default runtime on `main`**, and MLX is the optional one.

We are closing this PR without merging it because the version on `main` is a separate implementation with a different set of trade-offs, not because of the quality of yours — the ctypes struct layouts, the pinned commit and the sha256-pinned GGUF were all correct when we checked them against the upstream header and the Hugging Face files. What changed:

- **No build step.** `rizzo download` fetches the official prebuilt llama.cpp package for the machine (release `b11081`: CUDA, Vulkan, ROCm, SYCL, Metal, CPU), verified by sha256, so Windows and macOS are covered too (Windows tested, macOS not yet).
- **GGUF files from the model's authors** ([XHToken/Spark-X2.5-4B-GGUF](https://huggingface.co/XHToken/Spark-X2.5-4B-GGUF)), which were published in the meantime.
- **Devices are enumerated by ggml** instead of probing `/sys`, and prefix branches share KV cells in a unified cache instead of reserving `ctx × (batch + 1)`.
- `--device cuda` keeps working (the CLI rejected it on your branch when we tried it on an NVIDIA box).

Your contribution is credited in the [README](https://github.com/Rizzo-AI-Academy/rizzo-flow#credits).

One thing only you can do right now: everything on `main` was tested on a single Windows + NVIDIA machine (CUDA build and Vulkan build on the same card). **It has never run on an AMD GPU.** If you have ten minutes for your RX 7900 XTX:

```bash
git pull && uv sync --locked
uv run rizzo download        # should pick the Vulkan build; --runtime rocm for HIP
uv run rizzo devices
uv run rizzo decide examples/ticket.json
```

An issue with the output of `rizzo devices` and the timings — or with whatever breaks — would be the first real AMD report for this project. Thanks again.

**Comment by BiG86** [comment]
### First AMD report: RX 7900 XTX (gfx1100) — Vulkan works, `auto` picks the wrong build

Ran the four commands on `main` (`d34665b`), Linux x64, Radeon RX 7900 XTX. **Vulkan works and the
numbers are good.** Two things got in the way, and one of them contradicts the table in the README.

**Machine:** RX 7900 XTX (Navi 31, RADV), i9-12900KF, Linux x64, ROCm 7.2 installed
(`hip-runtime-amd`, `libamdhip64.so.5`), no NVIDIA GPU (`nvidia-smi` absent, no `/dev/nvidiactl`).

#### 1. What the AMD report should have been (Vulkan build)

`RIZZO_LLAMA_DIR=…-vulkan rizzo devices`:

```
Vulkan0 | gpu | Radeon RX 7900 XTX (RADV NAVI31) | 25.75 GB
CPU     | cpu | 12th Gen Intel(R) Core(TM) i9-12900KF
auto_selects: "Vulkan0"
```

`rizzo decide examples/ticket.json --device vulkan`, warm, median of 3:

| | cold | warm (median of 3) |
| --- | ---: | ---: |
| load | 3.7 s | 3.6 s |
| prefill | 305 ms | 87 ms |
| inference | 448 ms | 232 ms |
| total (4 questions) | 472 ms | 260 ms |
| batches / peak VRAM | 1 / 5.59 GiB | 1 / 5.59 GiB |

Answers on all four questions are identical to the CUDA/Windows numbers and identical on Vulkan,
CPU and the older branch — same argmax and same `p` to 4 decimals (`true` 0.9999, `access` 1.0000,
urgency 1.0000, `deadline_hours` `insufficient_evidence`). The pattern holds on AMD.

#### 2. Bug: `--runtime auto` installs the CUDA build on this AMD machine

```
$ uv run rizzo download
runtimes/llama-b11081-linux-x64-cuda        # not vulkan
models/Spark-X2.5-4B-GGUF/Spark-X2.5-4B-Q8_0.gguf

$ uv run rizzo devices
ggml_cuda_init: failed to initialize CUDA: no CUDA-capable device is detected
  "recommended": "cuda", "installed": ["cuda"], "auto_selects": "CPU"

$ uv run rizzo decide examples/ticket.json     # no flags, as the README shows
backend=cpu device=cpu
total 12 808 ms   (vs 260 ms on Vulkan — about 50× slower, ≈3.2 s per decision)
```

`recommended` stays `"cuda"` even with only the Vulkan build installed, so it is not a
"what is installed" report.

**Mechanism.** `llama_release.nvidia_driver()` decides from a *library name*:

```python
def nvidia_driver() -> bool:
    name = "nvcuda.dll" if sys.platform == "win32" else "libcuda.so.1"
    try:
        ctypes.CDLL(name)
    except OSError:
        return False
    return True
```

This box has a residual NVIDIA user-mode library — `/lib/x86_64-linux-gnu/libcuda.so.1 →
libcuda.so.580.178.04`, not owned by any dpkg package — left over from an earlier install. It
`dlopen`s fine, so `nvidia_driver()` returns `True` and `pick("auto")` returns `"cuda"` (line 192).
Loading the stub is not evidence of a GPU: `cuInit(0)` on the same machine returns **100**
(`CUDA_ERROR_NO_DEVICE`). Verified the library is the only cause: with the CUDA directory parked,
`auto_selects` becomes `Vulkan0` and `rizzo decide` runs on Vulkan at 256 ms.

**Why it matters beyond this one machine.** A stale `libcuda.so.1` is a realistic state (GPU
upgrade, leftover driver, container image), and the failure is silent: the only sign is one
`ggml_cuda_init` line on stderr, then a 50× slowdown with correct answers. It also contradicts the
README table, which tells AMD/Intel users that `auto` gives them `vulkan`:

> | Windows or Linux, AMD or Intel GPU | `vulkan` — uses the GPU driver you already have |

Documented behaviour and observed behaviour disagree on the first AMD machine that ran it.

**Suggested fix.** Probe for a device, not a library:

```python
def nvidia_driver() -> bool:
    """An NVIDIA GPU is usable: the driver initialises and reports at least one device."""
    try:
        cu = ctypes.CDLL("nvcuda.dll" if sys.platform == "win32" else "libcuda.so.1")
    except OSError:
        return False
    if cu.cuInit(0) != 0:           # 100 = CUDA_ERROR_NO_DEVICE
        return False
    count = ctypes.c_int(0)
    return cu.cuDeviceGetCount(ctypes.byref(count)) == 0 and count.value > 0
```

Independently of the heuristic, a GPU→CPU fallback that costs 50× deserves a visible warning — the
one time it was silent, it was because a leftover library had misled the code upstream of it.

#### 3. `--runtime rocm` installs but falls back to CPU, silently

```
$ uv run rizzo download --runtime rocm --only runtime
$ RIZZO_LLAMA_DIR=…-rocm uv run rizzo devices
devices: [ CPU ]                      # no GPU line, no warning
auto_selects: "CPU"
```

`libggml-hip.so` is in the tarball but never loads, and llama.cpp does not say so — even with
`RIZZO_LLAMA_LOG=1` the log ends at the RPC and CPU backends. The reason is a library version
mismatch:

```
libggml-hip.so: libhipblas.so.3: cannot open shared object file: No such file or directory
ldd: libhipblas.so.3 not found · librocblas.so.5 not found · libamdhip64.so.7 not found
```

The ROCm tarball is built for **ROCm 10.0**; this machine has **7.2** and provides
`libamdhip64.so.5`. That is arguably expected for an optional build, but `devices` showing a
healthy-looking CPU-only machine with no explanation is not: a HIP user has no way to tell a broken
backend from a machine without a GPU. A one-line note when a runtime directory contains a GPU
backend that failed to load would close the gap. Worth knowing whether ROCm 10 is the intended
floor, or whether the `-rocm-10.0` asset is being picked over a version-matched one.

#### Not tested

- macOS/Metal, Intel/SYCL — no hardware here.
- ROCm with a matching 10.x runtime — only the version-mismatch path above.
- Vulkan on the 1.7B and on `--quant bf16` / `q4_k_m`.

Happy to re-run any of this, or to test a patch for `nvidia_driver()` on this card.
